### PR TITLE
Ddqn

### PR DESCRIPTION
**Pull request by SergioMartin86**
- Added dueling DQN 
- Changed GFPT to use dualing networks

### CODE REVIEW

**Review by webepasc — APPROVED** [review]

File: source/modules/solver/agent/continuous/GFPT/GFPT._cpp [hunk 1]
```diff
@@ -105,16 +130,50 @@ void GFPT::initializeAgent()
...
+  // Calculating max_a A(s_i+1, a)
+  for (size_t j = 0; j < _problem->_stateVectorSize; j++) stateActionInput[j] = state[j];
+  for (size_t j = 0; j < _problem->_actionVectorSize; j++) stateActionInput[j + _problem->_stateVectorSize] = action[j];
+  aValAction = _aLearner->getEvaluation(stateActionInput)[0];
+
+  // Getting the Gaussian means of the actions given by the agent's policy
+  auto actionMeans = getActionMeans(state);
+
+  // Getting the Gaussian sigmas of the actions given by the agent's policy
+  auto actionSigmas = getActionSigmas(state);
```
Comment (wadaniel) [inline source/modules/solver/agent/continuous/GFPT/GFPT._cpp:148]: is it always diagonal?

File: source/modules/solver/agent/discrete/DDQN/DDQN._cpp [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+  // Creating state history indexes to choose from
+  std::vector<size_t> experienceReplayIndexes(_experienceReplayStates.size());
+  for (size_t i = 0; i < _experienceReplayStates.size() - 1; i++) experienceReplayIndexes[i] = i;
+
+  // Calculating cumulative Q*, for statistical purposes
+  _cumulativeQStar = 0.0;
+  _cumulativeQStarSquared = 0.0;
+
+  // Shuffling indexes to choose the mini batch from
+  std::shuffle(experienceReplayIndexes.begin(), experienceReplayIndexes.end(), *mt);
```
Comment (wadaniel) [inline source/modules/solver/agent/discrete/DDQN/DDQN._cpp:111]: I think we shall make this more efficient and do not shuffle 15k experiences if we only use 32 in the end

File: source/modules/solver/agent/discrete/DDQN/DDQN._cpp [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+  float sumExpQVal = 0.0;
+
+  // Iterating over all possible actions, checking their Q(s,a)
+  for (size_t i = 0; i < _problem->_possibleActions.size(); i++)
+  {
+    // Computing Q(s,a_i) = V(s) + A(s,a) - 1/n A(s,a')
+    float curQVal = vValState + aValAction[i] - aValAverage;
+
+    // Computing e^Q(s,a_i)
+    float expCurQVal = std::exp(curQVal);
```
Comment (wadaniel) [inline source/modules/solver/agent/discrete/DDQN/DDQN._cpp:264]: see dacer branch how to apply the logsum trick

Comment (wadaniel) [reply source/modules/solver/agent/discrete/DDQN/DDQN._cpp:264]: otherwise it can result in 0 probabilities everywhere if one q value is very large

**Pull request merged**